### PR TITLE
WalletManager refactoring part 1

### PR DESCRIPTION
**Pull request by bokobza**
* Moved the IWalletManager.CreateNewAccount method to the Wallet object and renamed it to AddNewAccount to better reflect this method's functionalities. 
I started with doing this but then instead of moving further along I worked on extracting HD-related methods to a separate class.
* Extracted all the methods relating to HD wallets' addresses and accounts to a separate class.
* Updated NStratis to the latest submodule

### CODE REVIEW

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs [hunk 1]
```diff
@@ -107,6 +107,23 @@ public IEnumerable<Script> GetAllPubKeysByCoinType(CoinType coinType)
                 yield return address.ScriptPubKey;
             }
         }
+
+        /// <summary>
+        /// Adds an account to the current wallet.
+        /// </summary>
+        /// <remarks>The name given to the account is of the form "account (i)" by default, where (i) is an incremental index starting at 0.
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: <remarks> in new line is easier no?

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs [hunk 2]
```diff
@@ -107,6 +107,23 @@ public IEnumerable<Script> GetAllPubKeysByCoinType(CoinType coinType)
...
+        /// According to BIP44, an account at index (i) can only be created when the account at index (i - 1) contains at least one transaction.
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: Maybe add to the comment that this method assumes that the BIP44 requirement of account i-1 has transactions

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs [hunk 3]
```diff
@@ -180,6 +197,57 @@ public HdAccount GetAccountByName(string accountName)
...
+        /// <param name="chainCode">The chain code for this wallet.</param>
+        /// <param name="network">The network for which this account will be created.</param>
+        /// <returns>A new hd account.</returns>
+        public HdAccount AddNewAccount(string password, string encryptedSeed, byte[] chainCode, Network network)
+        {
+            Guard.NotEmpty(password, nameof(password));
+            Guard.NotEmpty(encryptedSeed, nameof(encryptedSeed));
+            Guard.NotNull(chainCode, nameof(chainCode));
+
+            // Get the accounts for this type of coin.
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: Is the comment correct?

File: Stratis.Bitcoin/Features/Wallet/HdOperations.cs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        {
+            Guard.NotNull(privateKey, nameof(privateKey));
+            Guard.NotNull(chainCode, nameof(chainCode));
+
+            var accountHdPath = GetAccountHdPath(coinType, accountIndex);
+            return GetExtendedPublicKey(privateKey, chainCode, accountHdPath);
+        }
+
+        /// <summary>
+        /// Gets the extended public key for an account.
```
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/HdOperations.cs]: >Gets the extended public key for an HD path 

Is more accurate no?

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs (hunk 1 above)
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: you mean like this:
<remarks>
The name given to the account is of the form "account (i)" by default, where (i) is an incremental index starting at 0.
</remarks>

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs (hunk 2 above)
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: it doesn't assume it actually, the method called from this method checks whether the account has transactions or not.

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs (hunk 3 above)
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: nope, good spot.

File: Stratis.Bitcoin/Features/Wallet/HdOperations.cs (hunk 4 above)
Comment (bokobza) [inline Stratis.Bitcoin/Features/Wallet/HdOperations.cs]: yeah but it's the hd path of an account.
Maybe Gets the extended public key for an account corresponding to an HD path.

File: Stratis.Bitcoin/Features/Wallet/Wallet.cs (hunk 2 above)
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/Wallet.cs]: Ah ok cool so this is correct actually.

File: Stratis.Bitcoin/Features/Wallet/HdOperations.cs (hunk 4 above)
Comment (dangershony) [inline Stratis.Bitcoin/Features/Wallet/HdOperations.cs]: But any HD path can go in there unless there is a condition limiting the path to have an account

**Pull request merged**